Update for tyk-oss GA (#3036)
TykTechnologies/tyk-docs: tyk-docs/content/tyk-oss/setup-multiple-gateways.md

---
title: "Running multiple instances of Tyk Gateway"
date: 2023-08-01
tags: ["OSS", "Gateways", "Kubernetes"]
description: "Running multiple instances of Tyk Gateway"
menu:
  main:
    parent: "Tyk Helm Chart"
weight: 1
---

## Introduction

Running multiple instances of Tyk Gateway in Kubernetes can be tricky, as Tyk Dashboard (a licensed component) that helps synchronise API configurations across instances of gateways is missing.

> By default, Gateway stores API configurations at /mnt/tyk-gateway/apps inside the Gateway container. There are a a few challenges:

Reviewer: "are a a few" (doubled "a"). 

However I would rewrite the sentence to better signify that it's a consequence of the place where the API configuration are stored. Something like "This presents a few chalenges:"

Author: ```suggestion
By default, Gateway stores API configurations at /mnt/tyk-gateway/apps inside the Gateway container. This presents a few challenges:
```

- Multiple gateways do not share app configs

> - The configuration is not persistent. It got lost whenever pod restart

Reviewer: I think "It gets lost whenever a pod restarts" is the corect way of saying it.

Author: ```suggestion
- The configuration is not persistent. It gets lost whenever a pod restarts
```

> The same applies to Security Policies and middleware too which is stored at /mnt/tyk-gateway/policies and /mnt/tyk-gateway/middleware respectively.

Reviewer: The same applies to Security Policies and middleware**s** too which **are** stored at /mnt/tyk-gateway/policies and /mnt/tyk-gateway/middleware respectively.

Author: ```suggestion
The same applies to Security Policies and middlewares too which are stored at /mnt/tyk-gateway/policies and /mnt/tyk-gateway/middleware respectively.
```

> This can be solved by instantiating a Persistent Volume as shared storage for the gateway instances. As each gateway is reload, they would get the API configurations from the same storage, solving the synchronisation issue between gateways. Also, the storage is persistent and can be designed to be not impacted by cluster failure, your API configurations can be maintained after pod restart.

Reviewer 2: ```suggestion
This can be solved by instantiating a Persistent Volume as shared storage for the gateway instances. As each gateway is reloaded, they would get the API configurations from the same storage, solving the synchronisation issue between gateways. Also, the storage is persistent and can be designed to be not impacted by cluster failure, your API configurations can be maintained after pod restart.
```

Author: updated

> This can be solved by instantiating a Persistent Volume as shared storage for the gateway instances. As each gateway is reload, they would get the API configurations from the same storage, solving the synchronisation issue between gateways. Also, the storage is persistent and can be designed to be not impacted by cluster failure, your API configurations can be maintained after pod restart.

Reviewer: "to be not" is incorrect => "to not be"

 Also, the storage is persistent and can be designed **to not be** impacted by cluster failure, **therefore** your API configurations can be maintained after a pod restart.

Author: ```suggestion
This can be solved by instantiating a Persistent Volume as shared storage for the gateway instances. As each gateway is reloaded, they would get the API configurations from the same storage, solving the synchronisation issue between gateways. Also, the storage is persistent and can be designed to not be impacted by cluster failure, therefore your API configurations can be maintained after pod restart.
```

{{< img src="/img/diagrams/multiple-gateways.png" alt="multiple-gateways" >}}

 
## Steps

### 1. Create PersistentVolume and PersistentVolumeClaim

> Check your Kubernetes platform what kind of persistent storage is supported. Some platform supports Dynamic Storage Provisioning so you (or the Cluster Admin) do not need to create a PersistentVolume in advance. When an application requests storage through PersistentVolumeClaim, the corresponding persistent volume type would be provisioned. You can often specify type of storage through storageClassName.

Reviewer: Check your Kubernetes platform **to see which** kind of persistent storage is supported.
You can often specify **the** type of storage through storageClassName.

Author: ```suggestion
Check your Kubernetes platform to see which kind of persistent storage is supported. Some platform supports Dynamic Storage Provisioning so you (or the Cluster Admin) do not need to create a PersistentVolume in advance. When an application requests storage through PersistentVolumeClaim, the corresponding persistent volume type would be provisioned. You can often specify the type of storage through storageClassName.
```

Example: A PersistentVolumeClaim requesting Azure Disk storage on AKS:

Create `pvc-azure.yaml`:
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: tyk-app-claim
spec:
  storageClassName: managed-csi
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: tyk-policies-claim
spec:
  storageClassName: managed-csi
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
---
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: tyk-middleware-claim
spec:
  storageClassName: managed-csi
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
```

### 2. Install Multiple Instances of Tyk Gateways

> The following settings would generate 3 replicas of gateway that get Load Balanced behind the Azure Application Gateway. The gateway share the same persistent volume to store the APIs, policies, and middleware config.

Reviewer 2: ```suggestion
The following settings would generate 3 replicas of gateway that get Load Balanced behind the Azure Application Gateway. The gateway shares the same persistent volume to store the APIs, policies, and middleware config.
```

Author: updated

> The following settings would generate 3 replicas of gateway that get Load Balanced behind the Azure Application Gateway. The gateway share the same persistent volume to store the APIs, policies, and middleware config.

Reviewer: "midleware configs" as all the other elements of the enumeration are in plural form.

Author: ```suggestion
The following settings would generate 3 replicas of gateway that get Load Balanced behind the Azure Application Gateway. The gateway shares the same persistent volume to store the APIs, policies, and middleware configs.
```

  --set tyk-gateway.gateway.control.enabled=true \
  --set tyk-gateway.gateway.replicaCount=3 \
  --set "tyk-gateway.gateway.extraVolumes[0].name=tyk-app-storage" \
  --set "tyk-gateway.gateway.extraVolumes[0].persistentVolumeClaim.claimName=tyk-app-claim" \
  --set "tyk-gateway.gateway.extraVolumes[1].name=tyk-policies-storage" \
  --set "tyk-gateway.gateway.extraVolumes[1].persistentVolumeClaim.claimName=tyk-policies-claim" \
  --set "tyk-gateway.gateway.extraVolumes[2].name=tyk-middleware-storage" \
  --set "tyk-gateway.gateway.extraVolumes[2].persistentVolumeClaim.claimName=tyk-middleware-claim" \
  --set "tyk-gateway.gateway.extraVolumeMounts[0].name=tyk-app-storage" \
  --set "tyk-gateway.gateway.extraVolumeMounts[0].mountPath=/mnt/tyk-gateway/apps" \
  --set "tyk-gateway.gateway.extraVolumeMounts[1].name=tyk-policies-storage" \
  --set "tyk-gateway.gateway.extraVolumeMounts[1].mountPath=/mnt/tyk-gateway/policies" \
  --set "tyk-gateway.gateway.extraVolumeMounts[2].name=tyk-middleware-storage" \
  --set "tyk-gateway.gateway.extraVolumeMounts[2].mountPath=/mnt/tyk-gateway/middleware"
```

### 3. Use Tyk Operator to manage your APIs and Policies

> When the files in /apps or /policies folder is updated, you have to reload your gateways to pickup the change. Operator can help you to manage reload of the cluster groups automatically after each API or policies resource update.

Reviewer: you use "files" (plural) and "pickup the change" (singular) => it would be better to either use "file" or "changes" to keep the form constant.

Operator can help you manage the reload of the cluster groups automatically after each API or policies resource update.

Author: ```suggestion
When the files in /apps or /policies folder is updated, you have to reload your gateways to pickup the changes. Operator can help you to manage the reload of the cluster groups automatically after each API or policies resource update.
```
